control/prometheus.py

```diff
@@ -0,0 +1,164 @@
+#
+#  Copyright (c) 2021 International Business Machines
+#  All rights reserved.
+#
+#  SPDX-License-Identifier: LGPL-3.0-or-later
+#
+#  Authors: pcuzner@ibm.com
+#
+
+import os
+import logging
+import spdk.rpc as rpc
+
+from prometheus_client.core import REGISTRY, GaugeMetricFamily, CounterMetricFamily
+from prometheus_client import start_http_server
+
+
+def start_exporter(spdk_rpc_client, port):
+    """Start the prometheus exporter and register the NVMeOF collector"""
+    start_http_server(port)
+    REGISTRY.register(NVMeOFCollector(spdk_rpc_client))
+
+
+class NVMeOFCollector:
+    """Provide a prometheus endpoint for nvmeof gateway statistics"""
+
+    def __init__(self, spdk_rpc_client):
+        self.logger = logging.getLogger(__name__)
+        self.spdk_rpc_client = spdk_rpc_client
+        self.metric_prefix = "ceph_nvmeof"
+
+    def collect(self):
+        self.logger.debug("Processing prometheus scrape request")
+
+        spdk_version = rpc.spdk_get_version(self.spdk_rpc_client)
+        spdk = GaugeMetricFamily(
+            f"{self.metric_prefix}_spdk_metadata",
+            "SPDK Version information", 
+            labels=["version"])
+        spdk.add_metric([spdk_version.get("version", "Unknown")],1)
+        yield spdk
+
+        bdev_info = rpc.bdev.bdev_get_bdevs(self.spdk_rpc_client)
+        bdev_metadata = GaugeMetricFamily(
+            f"{self.metric_prefix}_bdev_metadata", 
+            "BDEV Metadata", 
+            labels=["bdev_name","pool_name","rbd_name"])
+        bdev_capacity = GaugeMetricFamily(
+            f"{self.metric_prefix}_bdev_capacity_bytes", 
+            "BDEV Capacity", 
+            labels=["bdev_name"])
+
+        for bdev in bdev_info:
+            bdev_size = bdev.get("block_size") * bdev.get("num_blocks")
+            bdev_capacity.add_metric([bdev.get("name")], bdev_size)
+            rbd_info = {}
+            try:
+                rbd_info = bdev["driver_specific"]["rbd"]
+            except KeyError:
+                print("no rbd information present?")
+            else:
+                bdev_metadata.add_metric([bdev.get("name"), rbd_info.get("pool_name"), rbd_info.get("rbd_name")], 1)
```

Comment:
Yes, you're right... but as usual we can't do little about our past mistakes 🤧 , except not making new ones 😛  Maybe just optionally enabling any metric family that scales linearly could partially fix the issue, while still providing enough flexibility in case these metrics are occasionally needed?

Reply:
I've added a parameter prometheus_bdev_pools to govern which pools to provide io stats for.

Personally I don't see the bdev/datastore performance stats as occassionally needed though.